### ISSUE TITLE
PUT 405 Method Not Allowed

### ISSUE THREAD

**Issue by rubinosoft** [mstilkerich/rcmcarddav#349]
Hi, 
when I try to create a contact, I get this error:

"PUT https://mydomain.com/server/addressbooks/username/sabre-vobject-xxxxxxxx-xxxx-xxxxx-xxxx-xxxxxxxxxxxx.vcf HTTP request was not successful (405 Method Not Allowed): 4.1.5 Sabre\DAV\Exception\MethodNotAllowed Creating new files in this collection is not supported"

Is there a way to resolve this issue?

Thank you.

**Comment by mstilkerich** [comment]
Hello,

  your server does not allow creation of new cards in the addressbook. Some things to check:

- Which CardDAV server are you using?
- Is the addressbook URL correct? Normally, I would expect one additional level under username for the addressbooks.
  - For example, can you see other contacts of the addressbook within rcmcarddav?
- If the URL is correct, does the server support readonly addressbooks, and if so, is this addressbook maybe configured as readonly?

**Comment by rubinosoft** [comment]
Thank you for your reply.

**Which CardDAV server are you using?**
I use the last version of sabreDAV (PHP)

**Is the addressbook URL correct? Normally, I would expect one additional level under username for the addressbooks.
For example, can you see other contacts of the addressbook within rcmcarddav?**
The URL is correct and using a Thunderbird addon (cardbook) it works fine

**If the URL is correct, does the server support readonly addressbooks, and if so, is this addressbook maybe configured as readonly?**
No, the addressbook are read/write

**Comment by mstilkerich** [comment]
Hello,

  Sabre/DAV is not a server that can be used standalone as far as I know, you would at least need some integration code.
I'm still wary because of the URL. If you look into [Sabre/DAV docs](https://sabre.io/dav/caldav-carddav-integration-guide/), you will find that addressbooks are one directory level under the username, like `/addressbooks/homer@example.org/book1`. This makes sense, as a user can have more than one addressbook.

Anyway, the server rejects the PUT request to create a new card. This works fine with other Sabre/DAV integrations (nextcloud, owncloud, Baïkal), so I really suspect an issue in your (assumed) integration code here. It would be plausible that Sabre/DAV rejects PUT requests in the addressbook home URI (`/addressbooks/homer@example.org`).

If you want, you can enable debugging in the plugin (inside plugins/carddav/config.inc.php, set
```php
$prefs['_GLOBAL']['loglevel'] = \Psr\Log\LogLevel::DEBUG;
$prefs['_GLOBAL']['loglevel_http'] = \Psr\Log\LogLevel::DEBUG;
```
and 

Then add the addressbook as a new addressbook and provide the generated log output in `logs/carddav.log` and `logs/carddav_http.log`. We then should be able to see what the server reports for addressbook home URI and the addressbooks. I cannot promise though we can resolve your issue with that information.

Finally, I looked in the Sabre/DAV source code: The 405 you receive is thrown for the addressbook home collection, i.e. the parent collection of the actual addressbooks. This backs my theory about the wrong URI.

**Comment by rubinosoft** [comment]
Hello,
we are using sabredav from 3 years with success; only after PHP 7 upgrade we get this issue. 
Before I provide logs, any ideas related?

**Comment by mstilkerich** [comment]
Hello, I have already provided you with my ideas. I’m also quite confident that the addressbook URL is wrong based on the error message you get and the circumstances under which sabre emits that error.

**Comment by rubinosoft** [comment]
Hello,
thank you very much for your support. I've resolved the issue: adding a / at the end of my server URL, it works!

Bye!
